src/mappings/referrer.ts

```diff
@@ -17,52 +18,88 @@ export function handleReferralCodeCreated(event: OnReferralCodeCreated): void {
+    const ACTION_ADD = 0
+    const ACTION_REMOVE = 1
+    const ACTION_UPDATE = 2
+
+    switch (event.params.action) {
+        case ACTION_ADD:
+            handleRefCodeAdd(event)
+            break
+        case ACTION_REMOVE:
+            handleRefCodeRemove(event)
+            break
+        case ACTION_UPDATE:
+            handleRefCodeUpdate(event)
+            break
+    }
+}
 
-            // uptick new referee daydata when an update or new referee is added
-            let newReferralReferees = newReferralCodeDayData.newReferees
-            newReferralReferees.push(event.params.addr.toHexString())
-            newReferralCodeDayData.newReferees = newReferralReferees
-            newReferralCodeDayData.save()
+function handleRefCodeAdd(event: OnReferralCodeUpserted) {
+    if (!event.params.newReferralCode) {
+        return
+    }
 
-            // update the referee code for the trader (sender)
-            trader.refereeCode = newReferralCode.id
-            trader.save()
+    const newRefCode = getReferralCode(event.params.newReferralCode)
+    if (!newRefCode) {
+        return
+    }
 
-            // update referee count for old code, if it exists
-            if (existingReferralCode) {
-                existingReferralCode.numReferees = existingReferralCode.numReferees.minus(BI_ONE)
-                existingReferralCode.save()
-            }
+    newRefCode.numReferees = newRefCode.numReferees.plus(BI_ONE)
+    newRefCode.save()
 
-            // update referee count for new code
-            newReferralCode.numReferees = newReferralCode.numReferees.plus(BI_ONE)
-            newReferralCode.save()
-        }
+    const trader = getOrCreateTrader(event.params.addr)
+    trader.refereeCode = event.params.newReferralCode
+    trader.save()
+}
+
+function handleRefCodeRemove(event: OnReferralCodeUpserted): void {
+    const oldRefCode = getReferralCode(event.params.oldReferralCode)
+    if (!oldRefCode) {
+        return
     }
+
+    oldRefCode.numReferees = oldRefCode.numReferees.minus(BI_ONE)
+    oldRefCode.save()
+
+    const trader = getOrCreateTrader(event.params.addr)
+    trader.refereeCode = null
+    trader.save()
+}
+
+function handleRefCodeUpdate(event: OnReferralCodeUpserted): void {
+    if (!event.params.newReferralCode || event.params.newReferralCode == event.params.oldReferralCode) {
+        return
+    }
+
+    const oldRefCode = getReferralCode(event.params.oldReferralCode)
+    if (!oldRefCode) {
+        return
+    }
+
+    const newRefCode = getReferralCode(event.params.newReferralCode)
+    if (!newRefCode) {
+        return
+    }
+
+    oldRefCode.numReferees = oldRefCode.numReferees.minus(BI_ONE)
+    oldRefCode.save()
+
+    newRefCode.numReferees = newRefCode.numReferees.plus(BI_ONE)
+    newRefCode.save()
+
+    const trader = getOrCreateTrader(event.params.addr)
+    trader.refereeCode = newRefCode.id
+    trader.save()
+
+    const newRefCodeDayData = getReferralCodeDayData(event, newRefCode.id)
```

Comment:
I guess we also need to do line 98 ~ 102 in `handleRefCodeAdd()` right? because case ADD will run through line 45 ~ 49 in the original code.

Reply:
Ah thx. I missed it. Fixed.